Lib/test/test_ssl.py

```diff
@@ -2471,6 +2470,39 @@ def test_check_hostname(self):
                                                 "check_hostname requires server_hostname"):
                         context.wrap_socket(s)
 
+        def test_check_hostname_idn(self):
+            if support.verbose:
+                sys.stdout.write("\n")
+
+            server_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+            server_context.load_cert_chain(IDN_CERT)
+
+            context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+            context.verify_mode = ssl.CERT_REQUIRED
+            context.check_hostname = True
+            context.load_verify_locations(IDN_CA)
+
+            # correct hostname should verify, when specified in several
+            # different ways
+            for server_hostname in ["pythön.example.org",
+                                    "xn--pythn-mua.example.org",
+                                    b"xn--pythn-mua.example.org"]:
+                server = ThreadedEchoServer(context=server_context, chatty=True)
+                with server:
+                    with context.wrap_socket(socket.socket(),
+                                             server_hostname=server_hostname) as s:
+                        s.connect((HOST, server.port))
+                        cert = s.getpeercert()
+                        self.assertTrue(cert, "Can't get peer certificate.")
+
+            # incorrect hostname should raise an exception
+            server = ThreadedEchoServer(context=server_context, chatty=True)
+            with server:
+                with context.wrap_socket(socket.socket(),
+                                         server_hostname="python.example.org") as s:
+                    with self.assertRaises(ssl.CertificateError):
+                        s.connect((HOST, server.port))
```

Comment:
We don't need a parametrised test: just have a method (or even a function!) that does all the setup and returns the wrapped socket.

As to "what difference it makes", the answer is that it makes it easier to isolate a test failure. For example, if this test fails with an assertion error that says `Can't get peer certificate`, what hostname failed? All of them? The only way to know is to re-run the tests under pdb and wait until the error is hit.

It also shadows failures: for example, if the test fails with one of the acceptable hostnames you don't know whether the functionality that fails on incorrect hostnames is still working, because that part of the test never runs.

Regardless, I wouldn't block the patch on this because it's a prevalent pattern: it's just a bad one. :grin: